### PR TITLE
Avoid AtxDb ATX lists

### PR DESCRIPTION
**Pull request by noamnelke**

### CODE REVIEW

**Comment by noamnelke** [comment]
✅Automation tests passed

File: activation/atxdb.go [hunk 1]
```diff
@@ -316,111 +307,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
...
+	AtxId   types.AtxId
+	LayerId types.LayerID
+}
+
 // addAtxToEpoch adds atx to epoch epochId
 // this function is not thread safe and needs to be called under a global lock
-func (db *ActivationDb) addAtxToEpoch(epochId types.EpochId, atx types.AtxId) error {
-	ids, err := db.atxs.Get(epochId.ToBytes())
-	var atxs []types.AtxId
+func (db *ActivationDb) updatePosAtxIfNeeded(atx *types.ActivationTx) error {
```
Comment (antonlerner) [inline activation/atxdb.go]: add UT for positive and negative cases for this funciton

File: activation/atxdb.go [hunk 2]
```diff
@@ -316,111 +307,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
...
+
+	err = db.atxs.Put([]byte(posAtxKey), idAndLayerBytes)
+	if err != nil {
+		return fmt.Errorf("failed to store posAtx ID and layer: %v", err)
+	}
+	return nil
+}
+
+func (db ActivationDb) getCurrentAtxIdAndLayer() (atxIdAndLayer, error) {
+	posAtxBytes, err := db.atxs.Get([]byte(posAtxKey))
```
Comment (antonlerner) [inline activation/atxdb.go]: change name frpm posAtxBytes

File: activation/atxdb.go [hunk 3]
```diff
@@ -316,111 +307,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
...
+	}
+
+	err = db.atxs.Put([]byte(posAtxKey), idAndLayerBytes)
+	if err != nil {
+		return fmt.Errorf("failed to store posAtx ID and layer: %v", err)
+	}
+	return nil
+}
+
+func (db ActivationDb) getCurrentAtxIdAndLayer() (atxIdAndLayer, error) {
```
Comment (antonlerner) [inline activation/atxdb.go]: What is this type/ function? why is that needed?

Comment (antonlerner) [reply activation/atxdb.go]: I think that this optimisation is no needed, just read the atx, it's only a few more bytes

Comment (noamnelke) [reply activation/atxdb.go]: The name of this function is unclear, I'll improve it: the "current ATX" is the ATX with the highest layer ID that we've ever seen. This ATX will be used as a positioning ATX.

Regarding the type, perhaps we can use the ATX header here instead of this new type. Didn't think introducing a type was a big deal.

Comment (noamnelke) [reply activation/atxdb.go]: Actually, we can't use the ATX header, since it doesn't contain the data required to calculate the ID, nor does it include the ID.

We could use the full ATX (incl. NIPST and view) but that's more than a few more bytes, and it also requires calculating the ID, which is expensive.

What's wrong with using a new type for this?

File: activation/atxdb.go (hunk 2 above)
Comment (noamnelke) [reply activation/atxdb.go]: done

File: activation/atxdb.go (hunk 3 above)
Comment (noamnelke) [reply activation/atxdb.go]: I've made naming improvements. I called the ATX with the highest layer, that we'll potentially use as a positioning ATX "top ATX".

I've also simplified and clarified some variable and function names as a result.

The code is clearer and easier to read now, IMO.

File: activation/atxdb.go (hunk 1 above)
Comment (noamnelke) [reply activation/atxdb.go]: Excellent point! Added.

File: activation/activation.go [hunk 4]
```diff
@@ -376,28 +370,23 @@ func (b *Builder) Load() *types.NIPSTChallenge {
 	return nil
 }
 
-func (b *Builder) GetPrevAtxId(node types.NodeId) (*types.AtxId, error) {
+func (b *Builder) GetPrevAtxId(node types.NodeId) (types.AtxId, error) {
 	//todo: make sure atx ids are ordered and valid
-	ids, err := b.db.GetNodeAtxIds(node)
+	id, err := b.db.GetNodeLastAtxId(node)
```
Comment (zalmen) [inline activation/activation.go:371]: Is the todo above still relevant?

File: activation/activation.go [hunk 5]
```diff
@@ -376,28 +370,23 @@ func (b *Builder) Load() *types.NIPSTChallenge {
...
-	return &ids[len(ids)-1], nil
+	return id, nil
 }
 
 // GetPositioningAtxId returns a randomly selected atx from the provided epoch epochId
 // it returns an error if epochs were not found in db
 func (b *Builder) GetPositioningAtxId(epochId types.EpochId) (*types.AtxId, error) {
 	//todo: make this on blocking until an atx is received
-	atxs, err := b.db.GetEpochAtxIds(epochId)
+	atxId, err := b.db.GetPosAtxId(epochId)
```
Comment (zalmen) [inline activation/activation.go:382]: This is the only place where you use `GetPosAtxId`, why not change it to return a pointer to the atx id?

File: activation/atxdb.go [hunk 6]
```diff
@@ -1,20 +1,18 @@
...
 	"github.com/spacemeshos/go-spacemesh/database"
 	"github.com/spacemeshos/go-spacemesh/log"
 	"github.com/spacemeshos/go-spacemesh/mesh"
 	"github.com/spacemeshos/go-spacemesh/types"
-	"sort"
 	"sync"
 )
 
 const CounterKey = 0xaaaa
+const topAtxKey = "topAtxKey"
```
Comment (zalmen) [inline activation/atxdb.go:16]: remove `CounterKey`

File: activation/atxdb.go [hunk 7]
```diff
@@ -256,33 +254,26 @@ func (db *ActivationDb) SyntacticallyValidateAtx(atx *types.ActivationTx) error
...
 func (db *ActivationDb) ContextuallyValidateAtx(atx *types.ActivationTxHeader) error {
 	if atx.PrevATXId != *types.EmptyAtxId {
-		prevAtxIds, err := db.GetNodeAtxIds(atx.NodeId)
+		lastAtx, err := db.GetNodeLastAtxId(atx.NodeId)
 		if err != nil {
 			db.log.WithFields(
-				log.String("atx_id", atx.ShortId()), log.String("node_id", atx.NodeId.Key)).
-				Error("could not fetch node ATXs: %v", err)
-			return fmt.Errorf("could not fetch node ATXs: %v", err)
+				log.String("atx_id", atx.ShortId()), log.String("node_id", atx.NodeId.ShortString())).
```
Comment (zalmen) [inline activation/atxdb.go]: use `log.AtxId` and `log.NodeId`

File: activation/atxdb.go [hunk 8]
```diff
@@ -256,33 +254,26 @@ func (db *ActivationDb) SyntacticallyValidateAtx(atx *types.ActivationTx) error
 // If a previous ATX is not referenced, it validates that indeed there's no previous known ATX for that miner ID.
 func (db *ActivationDb) ContextuallyValidateAtx(atx *types.ActivationTxHeader) error {
```
Comment (zalmen) [inline activation/atxdb.go:268]: Isn't validating the previous atx part of syntactic validity? 
Also, aren't we suppose to validate here the atx's sequence?

File: activation/atxdb.go [hunk 9]
```diff
@@ -366,111 +357,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
 	return common.BytesToUint32(val), nil
 }
 
+type atxIdAndLayer struct {
+	AtxId   types.AtxId
+	LayerId types.LayerID
+}
+
 // addAtxToEpoch adds atx to epoch epochId
```
Comment (zalmen) [inline activation/atxdb.go]: update the method's documentation

File: activation/atxdb.go [hunk 10]
```diff
@@ -366,111 +357,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
...
 // this function is not thread safe and needs to be called under a global lock
-func (db *ActivationDb) addAtxToEpoch(epochId types.EpochId, atx types.AtxId) error {
-	ids, err := db.atxs.Get(epochId.ToBytes())
-	var atxs []types.AtxId
+func (db *ActivationDb) updateTopAtxIfNeeded(atx *types.ActivationTx) error {
+	currentTopAtx, err := db.getTopAtx()
```
Comment (zalmen) [inline activation/atxdb.go:348]: don't you need to lock this section? What if you handle new ATXs in parallel? Is it thread safe?

File: activation/atxdb.go [hunk 11]
```diff
@@ -366,111 +357,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
...
-		atxs, err = decodeAtxIds(ids)
-		if err != nil {
-			return errors.New("could not get all atxs from database ")
-		}
+		return fmt.Errorf("failed to marshal top ATX: %v", err)
 	}
-	atxs = append(atxs, atx)
-	w, err := encodeAtxIds(atxs)
+
+	err = db.atxs.Put([]byte(topAtxKey), topAtxBytes)
```
Comment (zalmen) [inline activation/atxdb.go:365]: Keep `topAtxKey` as []byte instead of casting every time you use it

File: activation/atxdb.go [hunk 12]
```diff
@@ -366,111 +357,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
...
-					panic("inconsistent db")
-				}
-
-				return atx1.Sequence < atx2.Sequence
-			})
-		}
-	}
-
-	w, err := encodeAtxIds(atxs)
+	err := db.atxs.Put(key, atx.Id().Bytes())
```
Comment (zalmen) [inline activation/atxdb.go:392]: do you need to validate that it is indeed the latest atx that this node published?

File: activation/atxdb.go [hunk 13]
```diff
@@ -366,111 +357,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
...
-	return atxs, nil
+	return types.AtxId{Hash: common.BytesToHash(id)}, nil
 }
 
-// GetEpochAtxIds returns all atx id's created in this epoch as seen by this node
-func (db *ActivationDb) GetEpochAtxIds(epochId types.EpochId) ([]types.AtxId, error) {
-	db.RLock()
-	ids, err := db.atxs.Get(epochId.ToBytes())
-	db.RUnlock()
+// GetPosAtxId returns the best (highest layer id), currently known to this node, pos atx id
```
Comment (zalmen) [inline activation/atxdb.go:411]: let's add here a todo to consider another way to select the positioning atx once we introduce ticks

File: activation/atxdb.go [hunk 14]
```diff
@@ -366,111 +357,88 @@ func (db *ActivationDb) ActiveSetSize(epochId types.EpochId) (uint32, error) {
...
+	idAndLayer, err := db.getTopAtx()
 	if err != nil {
-		return nil, err
+		return *types.EmptyAtxId, err
 	}
-	atxs, err := decodeAtxIds(ids)
-	if err != nil {
-		return nil, err
+	if idAndLayer.LayerId.GetEpoch(db.LayersPerEpoch) != epochId {
+		return types.AtxId{}, fmt.Errorf("current posAtx (epoch %v) does not belong to the requested epoch (%v)",
```
Comment (zalmen) [inline activation/atxdb.go]: I don't entirely agree with how you handle errors in this method. I would expect that if you failed to use the top atx as a positioning atx, you'll find and return another candidate from the atx db. However, you should probably wait to see if this is indeed an issue (at least for the sake of not hiding bugs) before we add this functionality. 
Unless you disagree with me, please add a todo and file an issue.

**Review by zalmen — COMMENTED** [review]
I spotted only one new UT for rather a lot of new logic. Please validate that you have sufficient coverage for your new code with the existing UTs or add more.

File: activation/atxdb.go (hunk 6 above)
Comment (noamnelke) [reply activation/atxdb.go:16]: This is a great comment, only it's in the wrong PR... This is relevant in #1232, where I also failed to remove it and did it now, following your comment in https://github.com/spacemeshos/go-spacemesh/pull/1232/commits/66b271a0c0f30b503b655987aff667831c9982be

File: activation/atxdb.go (hunk 8 above)
Comment (noamnelke) [reply activation/atxdb.go:268]: No and no.

Syntactic: validating that we have access to the **referenced** previous ATX and that the sequence numbers match.

Contextual: validating that the referenced previous ATX matches the last ATX that we have **in our state** for that node ID.

Assuming that the contextual validation is done right before storing the ATX in question, this guarantees that there are no two ATXs that point to the same previous ATX. `processAtxMutex` in `ActivationDb` is there to ensure this assumption holds.

File: activation/atxdb.go (hunk 10 above)
Comment (noamnelke) [reply activation/atxdb.go:348]: The method's documentation says it's not thread safe and it should only be called under global lock. This part is not stale.

It's only called from `StoreAtx` which has a global lock from start to finish.

File: activation/atxdb.go (hunk 11 above)
Comment (noamnelke) [reply activation/atxdb.go:365]: I chose a string because it's immutable and can be a constant (unlike a byte array).

I could store it in a global variable instead and save me the casting, but I don't think it's that much better. If you insist, I'll do it.

File: activation/atxdb.go (hunk 12 above)
Comment (noamnelke) [reply activation/atxdb.go:392]: Contextual validation ensures that we store ATXs in the correct order.

It may have published another ATX, but this is the latest one in the DB.

File: activation/atxdb.go (hunk 13 above)
Comment (noamnelke) [reply activation/atxdb.go:411]: I kept this in mind when developing this part and the change should be minimal. I don't think a TODO to the code has much value because I won't remember to look for it.

What could be useful is to add a TODO to the ticks issue... or maybe that's what you meant?

I've created https://github.com/spacemeshos/go-spacemesh/issues/1257

File: activation/atxdb.go (hunk 14 above)
Comment (noamnelke) [reply activation/atxdb.go]: I partly disagree. Let me explain:

The top ATX is the most advanced (in terms of layers) ATX we have. When we build our own ATX we'll do it in the current epoch, never retroactively, so the options are: 

1. The top ATX belongs to the current epoch = 👍🏻
2. The top ATX belongs to an older epoch = 👎🏻

If option 2 materialized there's no point to look for another ATX, since the newest one is too old.

What we can (read: should) do in that case is to subscribe to a newer, relevant ATX being processed. There's a TODO for that one level up, here (credit for that TODO goes to @antonlerner):

https://github.com/spacemeshos/go-spacemesh/blob/bd59b3a505b77d9d76aeab60fba8e5a8a2cd053e/activation/activation.go#L382-L391

File: activation/activation.go (hunk 4 above)
Comment (noamnelke) [reply activation/activation.go:371]: no, removed

File: activation/activation.go (hunk 5 above)
Comment (noamnelke) [reply activation/activation.go:382]: done

File: activation/atxdb.go (hunk 7 above)
Comment (noamnelke) [reply activation/atxdb.go]: 👍🏻

File: activation/atxdb.go (hunk 9 above)
Comment (noamnelke) [reply activation/atxdb.go]: done

**Comment by noamnelke** [comment]
Regarding this comment: https://github.com/spacemeshos/go-spacemesh/pull/1239#pullrequestreview-270897082

I could only find a single case that wasn't covered, but I'm very glad I checked.

Added a test for this case now.

File: activation/atxdb.go (hunk 14 above)
Comment (noamnelke) [reply activation/atxdb.go]: Just realized that the documentation on that method I referenced is out of date... Pushed an update now:
https://github.com/spacemeshos/go-spacemesh/blob/40ac37c9a92460600006d1f2dfa421e548a29400/activation/activation.go#L381-L383

**Review by zalmen — APPROVED** [review]
